third_party/ibis/ibis_cloud_spanner/__init__.py

```diff
@@ -1,2 +1,220 @@
+        else:
+            db_value = database
+        db = self.instance.database(db_value)
+        tables = [tbl.table_id for tbl in db.list_tables()]
+
+        if like:
+            tables = [
+                table_name
+                for table_name in tables
+                if re.match(like, table_name) is not None
+            ]
+        return tables
+
+    def exists_table(self, name, database=None):
+
+        if database is None:
+            database = self.dataset_id
+
+        db_value = self.instance.database(database)
+        result = db_value.table(name).exists()
+        return result
+
+    def get_schema(self, table_id, database=None):
+        if database is None:
+            database = self.dataset_id
+        db_value = self.instance.database(database)
+        spanner_table = db_value.table(table_id)
+        return schema_from_spanner(spanner_table.schema)
+
+    def execute(
+        self,
+        expr: ir.Expr,
+        params: Mapping[ir.Scalar, Any] = None,
+        limit: str = "default",
+        **kwargs: Any,
+    ):
+        # Overwrite the execute() method to use Spanner to_pandas
+        kwargs.pop("timecontext", None)
+        query_ast = self.compiler.to_ast_ensure_limit(expr, limit, params=params)
+        sql = query_ast.compile()
+        self._log(sql)
+
+        schema = self.ast_schema(query_ast, **kwargs)
+
+        self._register_in_memory_tables(expr)
+        db = self.instance.database(self.dataset_id)
+
+        with db.snapshot() as snapshot:
+            result = pandas_df.to_pandas(snapshot, sql, query_parameters=None)
+
+        return result
+
+    def _execute(self, stmt, results=True, query_parameters=None):
+        db = self.instance.database(self.dataset_id)
+
+        with db.snapshot() as snapshot:
+            data_qry = pandas_df.to_pandas(snapshot, stmt, query_parameters)
+        return data_qry
+
+    def raw_sql(self, query: str, results=False, params=None):
+        db = self.instance.database(self.dataset_id)
+        with db.snapshot() as snapshot:
+            result = snapshot.execute_sql(query)
+        
+        return SpannerCursor(result)
+
+    def database(self, name=None):
+        if name is None and self.dataset is None:
+            raise ValueError(
+                "Unable to determine Cloud Spanner dataset. Call "
+                "client.database('my_dataset') or set_database('my_dataset') "
+                "to assign your client a dataset."
+            )
+        return self.database_class(name or self.dataset, self)
+
+    def set_database(self, name):
+        self.data_instance, self.dataset = self._parse_instance_and_dataset(name)
+
+    def dataset(self, database):
+        spanner_client = spanner.Client()
+        instance = spanner_client.instance(self.data_instance)
+        database = instance.database(database)
+
+    def exists_database(self, name):
+        return self.instance.database(name).exists()
+
+    def create_table():
```

Comment:
are these stubbed functions required?

Reply:
Yeah - the Ibis Backend function requires those methods to be present otherwise it'll throw an Exception